README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ineptR <a href="https://c-matos.github.io/ineptR/"><img src="man/figures/logo.png" align="right" height="132" /></a>

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version-last-release/ineptR?color=blue)](https://CRAN.R-project.org/package=ineptR)
[![R-CMD-check](https://github.com/c-matos/ineptR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/c-matos/ineptR/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/c-matos/ineptR/branch/main/graph/badge.svg)](https://app.codecov.io/gh/c-matos/ineptR?branch=main)
<!-- badges: end -->

## Overview

The goal of ineptR is to facilitate and automate data extraction from
Statistics Portugal (Instituto Nacional de Estatistica - INE, PT) with
R.  
It consists mainly of wrapper functions around the INE API, and some
helper functions to inspect metadata.  
One limitation of the INE API is that data extraction is limited to 40k
data points in each API call. For convenience, this package deals with
that limitation internally, and the user can get all the records in a
single `get_ine_data()` function call.

For a basic usage example, keep reading.  
For more in-depth examples, check `vignette("use_cases")`.

## Installation

You can install the development version of ineptR from
[GitHub](https://github.com/) with:

``` r
install.packages("ineptR") #Release version, from CRAN
# install.packages("devtools")
devtools::install_github("c-matos/ineptR") #Development version, from GitHub
```

## Usage

This is a basic workflow to inspect and get data from INE.

``` r
#Load package
library(ineptR)
```

1.  Navigate to the INE website and identify the desired indicador ID.

``` r
# Example: 
# Proportion of domestic budget funded by domestic taxes (%); Annual
# Available at: 
# https://www.ine.pt/xportal/xmain?xpid=INE&xpgid=ine_indicadores&indOcorrCod=0010003&contexto=bd&selTab=tab2&xlang=pt.
```

- Get the indicator ID from the indOcorrCod query parameter in the URL.
  In this case the ID is ***001003***

2.  Confirm that the indicator ID is valid:

``` r
#Check if indicator is valid
is_indicator_valid("0010003")
#> [1] TRUE
```

3.  Inspect metadata:

``` r
get_metadata("0010003")
#> $IndicadorCod
#> [1] "0010003"
#> 
#> $IndicadorNome
#> [1] "Percentagem do orçamento de Estado financiado por impostos cobrados internamente (%); Anual - Direção Geral do Orçamento (Ministério das Finanças)"
#> 
#> $Periodic
#> [1] "Anual"
#> 
#> $PrimeiroPeriodo
#> [1] "2010"
#> 
#> $UltimoPeriodo
#> [1] "2023"
#> 
#> $UnidadeMedida
#> [1] "Percentagem (%)"
#> 
#> $Potencia10
#> [1] "0"
#> 
#> $PrecisaoDecimal
#> [1] "2"
#> 
#> $Lingua
#> [1] "PT"
#> 
#> $DataUltimaAtualizacao
#> [1] "2022-12-28"
#> 
#> $DataExtracao
#> [1] "2023-08-17T21:01:49.507+01:00"
```

4.  Get information about the dimensions:

- How many dimensions are in this indicator and what are they?

``` r
get_dim_info("0010003")
#> # A tibble: 2 × 3
#>   dim_num abrv                            versao
#>   <chr>   <chr>                           <chr> 
#> 1 1       Período de referência dos dados XXXXX 
#> 2 2       Localização geográfica          03505
```

- What are the unique values for each dimension?

``` r
get_dim_values("0010003")
#> # A tibble: 15 × 7
#>    dim_num cat_id  categ_cod categ_dsg categ_ord categ_nivel value_id        
#>    <chr>   <chr>   <chr>     <chr>     <chr>     <chr>       <chr>           
#>  1 1       S7A2010 S7A2010   2010      20100101  1           Dim_Num1_S7A2010
#>  2 1       S7A2011 S7A2011   2011      20110101  1           Dim_Num1_S7A2011
#>  3 1       S7A2012 S7A2012   2012      20120101  1           Dim_Num1_S7A2012
#>  4 1       S7A2013 S7A2013   2013      20130101  1           Dim_Num1_S7A2013
#>  5 1       S7A2014 S7A2014   2014      20140101  1           Dim_Num1_S7A2014
#>  6 1       S7A2015 S7A2015   2015      20150101  1           Dim_Num1_S7A2015
#>  7 1       S7A2016 S7A2016   2016      20160101  1           Dim_Num1_S7A2016
#>  8 1       S7A2017 S7A2017   2017      20170101  1           Dim_Num1_S7A2017
#>  9 1       S7A2018 S7A2018   2018      20180101  1           Dim_Num1_S7A2018
#> 10 1       S7A2019 S7A2019   2019      20190101  1           Dim_Num1_S7A2019
#> 11 1       S7A2020 S7A2020   2020      20200101  1           Dim_Num1_S7A2020
#> 12 1       S7A2021 S7A2021   2021      20210101  1           Dim_Num1_S7A2021
#> 13 1       S7A2022 S7A2022   2022      20220101  1           Dim_Num1_S7A2022
#> 14 1       S7A2023 S7A2023   2023      20230101  1           Dim_Num1_S7A2023
#> 15 2       PT      PT        Portugal  1         1           Dim_Num2_PT
```

5.  Get the data:

``` r
get_ine_data("0010003")
#>    dim_1 geocod   geodsg valor
#> 1   2010     PT Portugal 58.31
#> 2   2011     PT Portugal 60.04
#> 3   2012     PT Portugal 60.53
#> 4   2013     PT Portugal 58.92
#> 5   2014     PT Portugal  62.1
#> 6   2015     PT Portugal 61.28
#> 7   2016     PT Portugal 64.61
#> 8   2017     PT Portugal 64.55
#> 9   2018     PT Portugal 67.41
#> 10  2019     PT Portugal 69.16
#> 11  2020     PT Portugal 58.01
#> 12  2021     PT Portugal 60.72
#> 13  2022     PT Portugal 64.18
#> 14  2023     PT Portugal 66.08
```


NEWS.md
# ineptR 0.2.1
* Added unit testing and codecov integration
* Corrected some edge cases that might result in errors

# ineptR 0.2.0

* Fixed issues to comply with the CRAN policy:
'Packages which use Internet resources should fail gracefully with an
informative message if the resource is not available or has changed (and
not give a check warning nor error).'

# ineptR 0.1.0

* First release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(ineptR)

test_check("ineptR")


tests/testthat/test-get_dim_info.R
test_that("Problems return NULL, with a message", {
  suppressMessages({
    expect_message(get_dim_info("invalid_id"),"The indicator code does not exist")
    expect_null(get_dim_info("invalid_id"))
  })
})

test_that("Indicator exists and call succeded", {
  suppressMessages({
    expect_type(get_dim_info("0011823"),"list")
    expect_s3_class(get_dim_info("0011823"),"data.frame")
    #expect_named(get_dim_info("0011823"))
    expect_true(all(c('dim_num','abrv','versao') %in% names(get_dim_info("0011823")))) #at least these columns must exist
  })
})

test_that("HTTP errors arre downgraded to messages", {
  suppressMessages({
    mock_404 <- function(req) {httr2::response(status_code = 404)}
    mock_503 <- function(req) {httr2::response(status_code = 503)}
    expect_message(httr2::with_mock(mock_404,get_dim_info("0011823")),"404")
    expect_message(httr2::with_mock(mock_503,get_dim_info("0011823")),"503")
  })
})


tests/testthat/test-get_ine_data.R
test_that("Problems return NULL, with a message", {
  suppressMessages({
    expect_message(get_ine_data("invalid_id"),"The indicator code does not exist")
    expect_null(get_ine_data("invalid_id"))
  })
})

test_that("Indicator exists and call succeded", {
  expect_type(get_ine_data("0008206", dim1 = "S7A1996", dim2 = c("11","111"), dim4 = c(1,19), dim5 = "TLES"),"list")
  expect_s3_class(get_ine_data("0011823"),"data.frame")
  expect_named(get_ine_data("0011823"))
  expect_true(all(c('dim_1','geocod','geodsg','valor') %in% names(get_ine_data("0011823")))) #at least these columns must exist
})

test_that("HTTP errors arre downgraded to messages", {
  suppressMessages({
    mock_404 <- function(req) {httr2::response(status_code = 404)}
    mock_503 <- function(req) {httr2::response(status_code = 503)}
    expect_message(httr2::with_mock(mock_404,get_ine_data("0011823")),"404")
    expect_message(httr2::with_mock(mock_503,get_ine_data("0011823")),"503")
  })
})

# test_that("Duration log succeeds", {
#   skip_on_cran()
#   expect_output(get_ine_data("0011823", expected.duration = T),"Output: 1 of")
# })

test_that("User input is correct", {
  expect_error(get_ine_data("0010003", Blah = c("1234"))) #Wrong dimension name
  expect_error(get_ine_data("0010003", dim1 = c("S7A2011"), dim2 = "PT", dim3 = "TLES")) #Impossible number of dimensions
})

test_that("Calculations work for request over the API 40k limit", {
  skip_on_cran()
  expect_gt(nrow(get_ine_data("0008206", dim1 = c("S7A1996"), dim3 = "T", dim5 = "TLES", max_cells = 2000)),5000) #over 40k values returned
  expect_gt(nrow(get_ine_data("0008206", dim1 = c("S7A2020"), dim2 = "PT", dim4 = "T", max_cells = 100)),150) #over 40k values returned
  expect_gt(nrow(get_ine_data("0001001", max_cells = 15)),30) #over 40k values returned
})


R/utils.R
#' Get raw indicator metadata
#'
#' @description
#' `r lifecycle::badge('experimental')`
#' `get_metadata_raw()` calls the metadata API.
#'
#' Be aware that this function will still return a value and not throw an error if the indicator does not exist.
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang Only "PT" implemented.
#'
#' @keywords internal
#' @return API response body as list. Returns NULL with an informative message if no internet connection, timeout or HTTP error.
#'
#' @seealso [ineptR::is_indicator_valid()] can be used to check if the indicator is valid before calling this function.
get_metadata_raw <- function(indicator, lang = "PT") {

  #Move the parameters to a list to splice later
  params <- list(
    varcd = indicator,
    lang = lang
  )
  baseurl <- "https://www.ine.pt/ine/json_indicador/pindicaMeta.jsp"

  #metadata_path <- sprintf("https://www.ine.pt/ine/json_indicador/pindicaMeta.jsp?varcd=%s&lang=%s", indicator, lang)
  #httr2::request(metadata_path) %>% httr2::req_perform() %>% httr2::resp_body_json() %>% magrittr::extract2(1)
  #jsonlite::fromJSON(txt = metadata_path, simplifyVector = F)[[1]] #Fails on macos-latest in R CMD check

  req <- httr2::request(base_url = baseurl) %>%
    httr2::req_url_query(!!!params) %>%
    httr2::req_user_agent("ineptR (https://c-matos.github.io/ineptR/)") %>%
    httr2::req_error(is_error = ~FALSE)  #Return message instead of error

  #Validates no internet connection, timeout errors and HTTP errors
  resp <- gracefully_fail(req)

  if (is.null(resp)) {
    return(invisible(NULL))
  }

  #Select only the desired part of the output
  metadata_raw <- resp %>%
    httr2::resp_body_json() %>%
    magrittr::extract2(1) # {\(x) x[[1]]}() --> equivalent, but compatible with base pipe

  #Return the resp to avoid duplicate calls. This is an internal function
  return(metadata_raw) #list(resp = resp, meta = metadata_raw)
}


#Calculate the number of dimensions in the indicator
#Will be used for validation purposes
calc_num_dims <- function(indicator, lang="PT") {
  metadata <- get_metadata_raw(indicator = indicator, lang = lang)

  if (is.null(metadata)) {
    return(invisible(NULL))
  }

  metadata %>%
    magrittr::use_series("Dimensoes") %>%
    magrittr::extract2(1) %>%
    length()
}

#Calculate the number of unique values in each dimension
calc_dims_length <- function(indicator, lang="PT") {
  dim_num <- NULL

  dim_values <- get_dim_values(indicator)

  if (is.null(dim_values)) {
    return(invisible(NULL))
  }

  dim_values %>%
    dplyr::group_by(dim_num) %>%
    dplyr::tally()
}


#Logic to deal with the limit of 40k cells per API call
get_api_urls <- function(indicator, max_cells=30000, lang="PT", ...) {
  dim_num <- cat_id <- n <- dim_1 <- NULL

  #unpack the ellipsis
  opt <- list(...)
  opt_names <- stringr::str_to_sentence(names(list(...))) #convert parameter names to sentence case
  names(opt) <- opt_names
  opt <- append(opt, values = c(op=2, varcd=indicator, lang=lang))

  #get number of dimensions
  num_dims <- calc_num_dims(indicator) #get number of dimensions

  if (is.null(num_dims)) {
    return(invisible(NULL))
  }

  #Get number of unique values in each dimension
  dims_len <- calc_dims_length(indicator)

  if (is.null(dims_len)) {
    return(invisible(NULL))
  }

  dim_values <- get_dim_values(indicator)

  if (is.null(dim_values)) {
    return(invisible(NULL))
  }


  # ------------------#
  # --- VALIDATION ---#
  # ------------------#

  # Check if parameters are named dimX, with max X being <= num_dims
  if (length(opt) - 3 > num_dims) {
    stop(sprintf("You are trying to extract more dimensions than are available for indicator: %s (%s passed, %s allowed)", indicator, length(opt)-3, num_dims))
  }

  #Check if any parameter is not in the form "dimN" (case insensitive)
  any_wrong_pattern <- !as.logical(prod(stringr::str_detect(opt_names, stringr::regex("^dim\\d$", ignore_case = T))))
  wrong_dimension_names <- opt_names[which(!stringr::str_detect(opt_names, stringr::regex("^dim\\d$", ignore_case = T)))]

  if (any_wrong_pattern) {
    stop(sprintf("All parameters should be in the form 'dimN'. Error at: %s", wrong_dimension_names))
  }

  #Check if user included dim1. If not, loop over all possible dim1 values
  if (!"Dim1" %in% opt_names) {
    my_dim1 <- dim_values %>%
      dplyr::filter(dim_num == 1) %>%
      dplyr::select(cat_id) %>%
      as.list() %>%
      magrittr::set_names("Dim1")
    opt <- append(x = opt, values = my_dim1)
  }

  #TODO
  ## Check if user input results in an extraction < max_rows

  #Get dimensions passed by the user
  user_dims <- opt_names %>%
    purrr::map_chr(~as.character(readr::parse_number(.))) %>%
    as.integer()

  #value to check if user input is already below max_rows
  rows_with_user_input <- dims_len %>%
    dplyr::filter(!dim_num %in% c(1, user_dims)) %>%
    dplyr::summarise(p = prod(n)) %>%
    as.integer()

  if (rows_with_user_input > max_cells) {
    #Calculate extra dimensions that need to be looped over, to ensure <= max_rows are extracted in each API call

    #Remaining dims, excluding dim1 an user specified dims
    remaining_dims <- dims_len %>%
      dplyr::filter(!dims_len$dim_num %in% c(1, user_dims))

    temp_out <- c()
    for (i in as.integer(remaining_dims$dim_num)) {
      mystat <- remaining_dims %>%
        dplyr::filter(!dim_num==i) %>%
        dplyr::summarise(p = prod(n)) %>%
        as.integer()

      temp_out <- c(temp_out, stats::setNames(mystat, i))
    }

    temp_out <- temp_out[temp_out<=max_cells]
    extra_dims_to_loop_over <- c(names(which.max(temp_out)))
    #Now get all the values of the desired dim, and add them to opt
    for (d in extra_dims_to_loop_over) {
      opt <- append(x = opt,
                    values = dim_values %>%
                      dplyr::filter(dim_num == d) %>%
                      dplyr::select(cat_id) %>%
                      as.list() %>%
                      magrittr::set_names(paste0("Dim",d)))
    }
  }

  #Base case, assuming everything is passed correctly
  #Generate urls with the parameters passed by the user, that are stored in opt

  #base ine api url (move to a previous location)
  baseurl <- "https://www.ine.pt/ine/json_indicador/pindica.jsp"

  #create the necessary urls for extraction
  urls <- tidyr::crossing(!!!opt) %>% #!!! is used to evaluate a list of expressions; !! would work for a vector
    purrr::pmap(list) %>%
    purrr::map_chr(~ httr::modify_url(baseurl, query=.x) )

  return(urls)
}


gracefully_fail <- function(request, path = NULL) {
  #Fails gracefully to comply with CRAN policy
  #path is used to save directly to disk. Not being used
  try_GET <- function(request) {
    tryCatch(
      resp <- request %>%
        httr2::req_perform(path = path),
      error = function(e) conditionMessage(e),
      warning = function(w) conditionMessage(w)
    )
  }
  is_response <- function(x) {
    class(x) == "httr2_response"
  }

  # First check internet connection
  if (!curl::has_internet()) {
    message("No internet connection.")
    return(invisible(NULL))
  }
  # Then try for timeout problems
  resp <- try_GET(request)
  if (!is_response(resp)) {
    message(resp)
    return(invisible(NULL))
  }
  # Then stop if status > 400
  if (httr2::resp_status(resp)>=400) {
    message(paste(httr2::resp_status(resp), httr2::resp_status_desc(resp), sep = " - "))
    return(invisible(NULL))
  }
  resp
}


R/get_dim_info.R
#' Get information about the dimensions of a given indicator
#'
#' @description
#' `r lifecycle::badge('experimental')` \cr
#' Get information about the dimensions of a given indicator
#'
#'
#' @details
#' If the indicator is not valid, returns "(PT) O codigo do indicador nao existe. / (EN) The indicator code does not exist." \cr
#' Calling `is_indicator_valid()` before using this function is recommended.
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang One of "PT" or "EN". Default is "PT".
#' @return A data frame with dim_num (dimension number), abrv (dimension description) and versao (dimension version) for the selected indicator.
#'         A fourth column 'nota_dsg' is present for some indicators, with additional notes about the dimensions.
#' @importFrom magrittr %>%
#' @export
#'
#' @examples
#' get_dim_info("0011823", lang = "EN")

get_dim_info <- function(indicator, lang="PT") {
  value <- NULL
  valid_ind <- is_indicator_valid(indicator)
  if (is.null(valid_ind)) {
    return(invisible(NULL))
  }

  if (valid_ind) {

    metadata <- get_metadata_raw(indicator = indicator, lang = lang)

    if (is.null(metadata)) {
      return(invisible(NULL))
    }

    metadata %>%
      magrittr::extract2("Dimensoes") %>%
      magrittr::extract2("Descricao_Dim") %>%
      tibble::as_tibble_col() %>%
      tidyr::unnest_wider(col = value)
  } else {
    #message("(PT) O código do indicador não existe. / (EN) The indicator code does not exist.")
    message("(PT) O c\u00F3digo do indicador n\u00E3o existe. / (EN) The indicator code does not exist.")
    return(invisible(NULL))
    # metadata <- get_metadata_raw(indicator = indicator, lang = lang)
    #
    # if (is.null(metadata)) {
    #   return(invisible(NULL))
    # }
    #
    # metadata %>%
    #   magrittr::extract2("Sucesso") %>%
    #   magrittr::use_series("Falso") %>%
    #   magrittr::extract2(1) %>%
    #   magrittr::use_series("Msg")
  }
}


R/get_metadata.R
#' Get indicator metadata
#'
#' @description
#' `r lifecycle::badge('experimental')` \cr
#' Get metadata for the selected indicator from the INE API
#'
#' Be aware that this function will still return a value and not throw an error if the indicator does not exist.
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang One of "PT" or "EN". Default is "PT".
#'
#' @return API response body as list.
#'
#' @seealso [ineptR::is_indicator_valid()] can be used to check if the indicator is valid before calling this function.
#'
#' @export
#' @examples
#' get_metadata("0011823")
#'
get_metadata <- function(indicator, lang = "PT") {
  temp_metadata <- get_metadata_raw(indicator = indicator, lang = lang)
  if (is.null(temp_metadata)) {
    return(invisible(NULL))
  }
  temp_metadata["Dimensoes"] <- NULL
  temp_metadata["Sucesso"] <- NULL
  if (length(temp_metadata)==0) {
    #message("(PT) O código do indicador não existe. / (EN) The indicator code does not exist.")
    message("(PT) O c\u00F3digo do indicador n\u00E3o existe. / (EN) The indicator code does not exist.")
    return(invisible(NULL))
  }
  return(temp_metadata)
}


R/get_dim_values.R
#' Get the set of possible values for all dimensions of the indicator
#'
#' @description
#' `r lifecycle::badge('experimental')` \cr
#' Get the set of possible values for all dimensions of the indicator
#'
#'
#' @details
#' If the indicator is not valid, returns "(PT) O codigo do indicador nao existe. / (EN) The indicator code does not exist." \cr
#' Calling `is_indicator_valid()` before using this function is recommended.
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang One of "PT" or "EN". Default is "PT".
#' @return A tidy data frame with dim_num (dimension number), cat_id (dimension description),
#'         categ_cod, categ_dsg (Category description), categ_ord (order in which the category appears in the website),
#'         categ_nivel (hierarchical order) and value_id (id row for this data frame) for the selected indicator.
#' @importFrom magrittr %>%
#' @export
#'
#' @examples
#' get_dim_values("0011823")
get_dim_values <- function(indicator, lang = "PT") {
  value <- NULL
  valid_ind <- is_indicator_valid(indicator)

  if (is.null(valid_ind)) {
    return(invisible(NULL))
  }

  if (valid_ind) {

    metadata <- get_metadata_raw(indicator = indicator, lang = lang)

    if (is.null(metadata)) {
      return(invisible(NULL))
    }

    metadata %>%
      magrittr::extract2("Dimensoes") %>%
      magrittr::extract2("Categoria_Dim") %>%
      tibble::as_tibble_col() %>%
      tidyr::unnest_longer(col = value) %>%
      tidyr::unnest(value) %>%
      tidyr::unnest_wider(value)
  } else {
    message("(PT) O c\u00F3digo do indicador n\u00E3o existe. / (EN) The indicator code does not exist.")
    return(invisible(NULL))
  }
}


R/is_indicator_valid.R
#' Check if indicator exists
#'
#' @description
#' `r lifecycle::badge('experimental')` \cr
#' Check if the indicator exists and is callable via the INE API. \cr
#' Indicator ID can be found in the url under the 'indOcorrCod' query parameter when browsing the website. \cr
#' Example: https://www.ine.pt/xportal/xmain?xpid=INE&xpgid=ine_indicadores&indOcorrCod=0010003&contexto=bd&selTab=tab2
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang One of "PT" or "EN". Default is "PT".
#'
#' @return TRUE if indicator exists, otherwise FALSE.
#' @export
#'
#' @examples
#' is_indicator_valid("0011823")
is_indicator_valid <- function(indicator, lang="PT") {
  metadata <- get_metadata_raw(indicator = indicator, lang = lang)
  if (is.null(metadata)) {
    return(invisible(NULL))
  }
  return(!is.null(metadata$Sucesso$Verdadeiro))
}


R/get_ine_data.R

#' Get data from the INE API
#'
#' @description
#' `r lifecycle::badge('experimental')` \cr
#' Return tidy data frames from Instituto Nacional de Estatística API
#'
#' @details
#' Extraction time can be very high (many hours) for indicators with many dimensions and/or many unique values. E.g. Indicator "0008206 - Deaths by Place of residence, Sex, Age group and Death cause, Annual" will take many hours to extract.
#'
#'
#' @param indicator INE indicator ID as a 7 character string. Example: "0010003".
#' @param lang One of "PT" or "EN". Default is "PT".
#' @param expected.duration If TRUE, prints the expected remaining duration in the console.
#' @param max_cells Integer smaller than or equal to 40000, the maximum number of cells allowed in each API call.
#'                  Default value of 30000
#' @param ... Values for each Dimension of the indicator.
#'            Each parameter should be in the form dimN, with N one of \{1, ..., Nº of dimensions\}.
#'            If one of the dimensions is not included, output includes all values for that dimension.
#'
#' @return tidy data frame for the given indicator.
#'
#' @importFrom magrittr %T>%
#' @importFrom utils txtProgressBar
#' @importFrom utils setTxtProgressBar
#'
#' @seealso [ineptR::get_dim_values()] can be used to identify the values to be passed to *dimN* parameters, in the variable categ_cod.
#'  See \href{https://c-matos.github.io/ineptR/articles/use_cases.html}{this vignette} for further clarification on how to obtain only a subset of the data.
#'
#' @export
#'
#' @examples
#' #A simple example. Returns the complete dataset.
#' get_ine_data("0011823")
#' # A more complex example. Returns a subset of the dataset.
#' \donttest{
#' get_ine_data("0008206", dim1 = "S7A1996", dim2 = c("11","111"),
#'              dim4 = c(1,19), dim5 = "TLES")
#' }
get_ine_data <- function(indicator, lang="PT", expected.duration=FALSE, max_cells = 30000, ...) {
  . <- dim_1 <- NULL
  #get the urls
  myurls <- get_api_urls(indicator, max_cells, ...)

  if (is.null(myurls)) {
    return(invisible(NULL))
  }

  #TODO:
  ## Print the number of records extracted in each call

  #start counters to estimate extraction duration
  if (expected.duration) {
    progressr::handlers(global = T)
    progressr::handlers("cli")
    p <- progressr::progressor(along = myurls)
  }

  #!!!!!!!
  ret_data <- myurls %>%
    tibble::as_tibble() %>%
    dplyr::mutate(id = dplyr::row_number()) %>%
    purrr::pmap_dfr(function(...) {
      current <- tibble::tibble(...) # do stuff and access content from current row with "current"

        if (expected.duration) {
          p(sprintf("Extracting %g", current$id))
        }

      e <- new.env()
      req <- current$value %>%
        purrr::map(~httr2::request(base_url = .x) %>%
                     httr2::req_user_agent("ineptR (https://c-matos.github.io/ineptR/)") %>%
                     httr2::req_error(is_error = ~FALSE))

      resp <- req %>% purrr::map(gracefully_fail)

      if (is.null(resp)) {
        return(invisible(NULL))
      }

      resp[[1]] %>%
        httr2::resp_body_json() %>%
        magrittr::extract2(1) %>%
        magrittr::use_series("Dados") %T>%
        {assign("temp_dim1", names(.), envir = e)} %>%
        magrittr::extract2(1) %>%
        purrr::map_dfr(data.frame) %>%
        dplyr::mutate(dim_1 = e$temp_dim1)
    }) %>%
    dplyr::relocate(dim_1, .before=1)

  if (expected.duration) {
    progressr::handlers(global = F)
  }
  return(ret_data)
}
